File: algorithm_library/python/hash/MultisetHash.py
from random import randint
class MultisetHash:
  # 比較したい配列を引数に渡す
  def __init__(self,A: list, B: list) -> None:
    self.mod = (1<<61)-1
    self.hash = dict()
    for a in A: self.hash[a] = randint(1,self.mod-1)
    for b in B: self.hash[b] = randint(1,self.mod-1)
    # ハッシュを取る
    self.hashA,self.hashB = [0],[0]
    for a in A: 
      self.hashA.append((self.hashA[-1] + self.hash[a]) %self.mod)
    for b in B: 
      self.hashB.append((self.hashB[-1] + self.hash[b]) %self.mod)
    
  # 1-index(A_1, A_2, ...)
  # A[al:ar+1)==B[bl:br+1) を返す
  def isEqual(self,al,ar,bl,br):
    return (self.hashA[ar] - self.hashA[al-1])%self.mod == (self.hashB[br] - self.hashB[bl-1])%self.mod

File: algorithm_library/cpp-from-py/hash/MultisetHash.cpp
#include <bits/stdc++.h>
using namespace std;

template <class T>
struct MultisetHash {
    static constexpr unsigned long long mod = (1ULL << 61) - 1;
    unordered_map<T, unsigned long long> hash_value;
    vector<unsigned long long> hashA, hashB;
    mt19937_64 rng;

    MultisetHash(const vector<T>& A, const vector<T>& B) : rng(chrono::steady_clock::now().time_since_epoch().count()) {
        for (const auto& a : A) hash_value[a] = randint();
        for (const auto& b : B) hash_value[b] = randint();
        hashA = {0};
        hashB = {0};
        for (const auto& a : A) hashA.push_back((hashA.back() + hash_value[a]) % mod);
        for (const auto& b : B) hashB.push_back((hashB.back() + hash_value[b]) % mod);
    }

    unsigned long long randint() {
        return uniform_int_distribution<unsigned long long>(1, mod - 1)(rng);
    }

    bool isEqual(int al, int ar, int bl, int br) const {
        return (hashA[ar] + mod - hashA[al - 1]) % mod == (hashB[br] + mod - hashB[bl - 1]) % mod;
    }
};
